src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.netcoreapp.cs

```diff
@@ -0,0 +1,54 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace System.Diagnostics
+{
+    partial class Activity
+    {
+        /// <summary>
+        /// Returns high resolution (~1 usec) current UTC DateTime.
+        /// </summary>
+        internal static DateTime GetUtcNow()
+        {
+            // .NET Core CLR gives accurate UtcNow
+            return DateTime.UtcNow;
+        }
+
+        private unsafe static SpanAction<char, long> s_generateRootId = (span, rootId) =>
+        {
+            span[0] = '|';
+            char format = 'x';
+            span = span.Slice(1);
+            // It is important that the part that changes frequently be first, because
+            // many sampling functions don't sample from the high entropy part of their hash function.
+            // This makes sampling based on this produce poor samples.
+            rootId.TryFormat(span, out int charsWritten, new ReadOnlySpan<char>(&format, 1));
+
+            Debug.Assert(span.Length == charsWritten + s_uniqSuffix.Length);
+
+            span = span.Slice(charsWritten);
+
+            s_uniqSuffix.AsSpan().CopyTo(span);
+        };
+
+        private unsafe string GenerateRootId()
+        {
+            var rootId = Interlocked.Increment(ref s_currentRootId);
+            var length = 1 + CountHexDigits((ulong)rootId) + s_uniqSuffix.Length;
+
+            return string.Create(length, rootId, s_generateRootId);
+        }
```

Comment:
This seems like unnecessary complexity to avoid a copy of a few bytes. How about just something like:
```C#
private static string GenerateRootId()
{
    Span<char> result = stackalloc char[1 + 16 + s_uniqSuffix.Length]; // max length needed
    result[0] = '|';
    bool formatted = Interlocked.Increment(ref s_currentRootId).TryFormat(result.Slice(1), out int charsWritten, "x");
    Debug.Assert(formatted);
    s_uniqSuffix.AsSpan().CopyTo(result.Slice(1 + charsWritten));
    return new string(result.Slice(0, 1 + charsWritten + s_uniqSuffix.Length));
}

Reply:
This is a bit simpler and more portable; will go with this 😄

Reply:
Changed

Reply:
`Interlocked.Increment`

Why is this needed?

Reply:
Unique id for each new RootId created; can be called concurrently